proposals/1681/naming_schemas.md

```diff
@@ -0,0 +1,152 @@
+
+Although one drawback with this type of schema is that the casing must have its own rule as to what fields it applies to; in the example, the `namespace` does not need the rule while the `name` does.
+
+Another potential json schema could look like:
+
+[Generic Example]
+```json
+{
+    "integration_name": {
+        "type": "Job" | "Dataset",
+        "namespace": "^(?<uri>[A-Za-z]+://)(?<auth>[A-Za-z0-9]*)\.(?<host>[A-Za-z0-9]+):(?<port>[0-9]{1,6})$",
+        "name": "^(?<database>[A-Z]+)\.(?<schema>[A-Z]+)\.(?<table>[A-Z]+)$",
+        "unique_name": "$namespace/$name"
+  }
+}
+```
+
+[Snowflake Example]
+```json
+{
+    "snowflake": {
+        "type": "Dataset",
+        "namespace": "^(?<scheme>snowflake://)(?<auth>[A-Za-z0-9]{9}\.[a-z]{2}-[a-z]{4,6}-[0-9]{1})$",
+        "name": "^(?<database>[A-Z]+)\.(?<schema>[A-Z]+)\.(?<table>[A-Z]+)$",
+        "unique_name": "$namespace/$name"
+  }
+}
+```
+
+In the above outline, regular expressions replace the `namespace` and `name` lists. This also negates the need for examples, as the required input is self-evident (or at least as evident as regexes ever are). The regular expressions in each of `namespace` and `name` have named matching groups for each componenet, and are themselves referenced in whole by `unique_name`, which provides the rule for combining these expressions. Finally, the two fields to determine casing are removed, as the regular expression conveys the necessary casing rules.
+
+A major drawback of using regular expressions here is the rigidity of the expression: real-world input is messy, and restrictive expressions will likely break on legitimate cases. 
+
+Both of the above examples also do not allow for multiple correct `unique_names`, for instance, if a warehouse accepts a URI both with and without a region, both of the above would only accept cases without the region. This may be solved by either having multiple files for these cases, or, in the regex example, have a list of valid regexes in `namespace` or `name`. To ensure a valid `name` and `namespace` is always supplied, additions to the `Dataset` and `Job` classes in the client should be added in the form of validators. An additional naming format fields could be added which would represent the valid syntax of the `name` and `namespace`, as well as the valid way to combine these fields into a unique name. The validator would verify that the `Dataset` or `Job` instance's `name` and `namespace` fields conform to this given naming format. The naming format fields would come from the new naming schemas, and should be added to `Dataset` or `Job` instances at runtime, when the appropriate integration is known.
+
+As an example, the `Dataset` class would be modified to use regular expressions to check the `name` and `namespace`:
+
+```python
+@attr.s
+class Dataset(RedactMixin):
```

Comment:
I think we should take the advantage of possibly having strong typing and provide `SnowflakeDataset`, `PostgresDataset` etc that inherit from `Dataset`. 

This is why I wanted that naming be specified as JSON schema, to generate those classes instead of having one class that mixes validation, action and pure data structure - which is one of the bigger failings of Airflow IMO 🙂